### PR TITLE
Change repo structure

### PR DESCRIPTION
**Pull request by dominiquesydow**
## Description

Restructure data and notebook folder/file structure, so that each talktorial is self-contained in one folder (convenient wrt CI) with the numbering scheme 001, 002, ect.

Put all talktorial folders in `teachopencadd` folder which will become the module when applying the cookiecutter template (in another PR).

Example:
```
teachopencadd/
    001_chembl/
        talktorial.ipynb
        data/
        images/
        README
    002_bla/
        talktorial.ipynb
        data/
        images/
        README
    …
    010_/
    011_/
    012_/
    __init__.py
```

## ToDos
- [x]  Set up template folder `teachopencadd/000_template` and __get feedback before applying it to all talktorials__.
- [x]  Set up `teachopencadd/README.md` with new folder names and __get feedback before applying it to all talktorials__: [Proposed talktorial folder names](https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/README.md)
- [x] Move talktorials to new folders; per talktorial do:
  - Move notebook and rename to `talktorial.ipynb`.
  - Move data (this included IO data files but also plots generated in the talktorial).
  - Move images.
  - Update IO file paths in talktorial.
  - Remove talktorial numbers from notebook header (e.g. "# Talktorial 1"), since it is enough to have the index once stated in the talktorial folder name. 
  - Add note for talktorials 001-010: Reference to T1-T10 and the first TeachOpenCADD paper.

## Questions

- [x] Can we drop the introductory_notebooks 2-4? I think we cover all these concepts in notebooks T1+ with more explanation for both theory/code
- [x] We could revamp and keep introductory notebook 1 for a quick start to Python/Jupyter, enriched by the first notebook from [AI in Medicine](https://github.com/volkamerlab/ai_in_medicine/blob/master/week1_session1_grundkonzepte.ipynb). What do you think? 
- [x]  What do you think of the proposed folder naming scheme in `teachopencadd/README.md`? (see ToDos point 2)
- [x]  What do you think of the proposed topic split/names for T11/12 in `teachopencadd/README.md`? (see ToDos point 2)
- [x] Do this after refactoring for all notebooks (some notebooks are missing at the moment because they will come from an existing large notebook)?
  - Fill talktorials' [`README`](https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/000_template/README.md): Automatically (?) fetch notebook title, "Aims of this talktorial", and "Learning goals"?
  - Fill talktorials' [`data/README`](https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/000_template/data/README.md): Automatically (?) list here description of important IO/files (write `data/README` in notebook)?
  - Add citation header to all notebooks (see (issue)[https://github.com/volkamerlab/TeachOpenCADD/issues/25])?

## Status

- [x] Ready to go

### CODE REVIEW

**Comment by dominiquesydow** [comment]
@AndreaVolkamer, @jaimergp, could you please take a look at (i) the template talktorial folder (including READMEs) and (ii) the proposed talktorial folder names? Please also take a look at the questions in the PR text. 

After agreeing on a scheme, I will apply the template to all notebooks.

**Comment by jaimergp** [comment]
Yay, it's happening! I'm on it!

**Comment by jaimergp** [comment]
```
teachopencadd/
    001_chembl/
        talktorial.ipynb
        data/
        images/
        README
    002_bla/
        talktorial.ipynb
        data/
        images/
        README
    …
    010_/
    011_/
    012_/
__init__.py
```

^ Here the last `__init__.py` should be indented right? Like this:

```
    011_/
    012_/
    __init__.py
```


***

> Use foldername structure `index_biologicalentity_topic`

I am not sure about categorizing stuff right away in the folder name. If we go this way, we should define a glossary of accepted biological entities before the list grows too big, but that might be too constraining and in the end we will have a category system that doesn't work or it's too vague.

In my opinion, we should use titles that describe the contents precisely but succinctly (three, four words). Then, in the future, we ca use other means to provide collections of notebooks that make sense together (e.g. curated markdown files, symlinked directories, website sections...). 

***

> Questions:
> Can we drop the introductory_notebooks 2-4? I think we cover all these concepts in notebooks T1+ with more explanation for both theory/code

Agreed.

> We could revamp and keep introductory notebook 1 for a quick start to Python/Jupyter, enriched by the first notebook from AI in Medicine. What do you think?

100% in.

**Comment by dominiquesydow** [comment]
> I am not sure about categorizing stuff right away in the folder name. If we go this way, we should define a glossary of accepted biological entities before the list grows too big, but that might be too constraining and in the end we will have a category system that doesn't work or it's too vague.

I was also unsure about this, since we plan on grouping the talktorials into topic collections anyways.

> In my opinion, we should use titles that describe the contents precisely but succinctly (three, four words). Then, in the future, we ca use other means to provide collections of notebooks that make sense together (e.g. curated markdown files, symlinked directories, website sections...).

Will suggest new topic/folder names in a bit.

**Comment by dominiquesydow** [comment]
> Here the last __init__.py should be indented right? Like this:

Agreed.

**Comment by AndreaVolkamer** [comment]
@dominiquesydow  and @jaimergp  just thought about the talktorial names, maybe we could come up with a more categorical name?
Like the first ones operate all on compounds, so they could all start with compound_xxx, eg. change 01_chembl_comound_query to 01_compound_chembl_query and so on 
Not sure if we can fit all, but maybe we find a good scheme?

**Comment by dominiquesydow** [comment]
@jaimergp, @AndreaVolkamer, could you please let me know how you feel about the new talktorial index names? 
https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/README.md

File: teachopencadd/README.md [hunk 1]
```diff
@@ -3,19 +3,19 @@
...
+- `004_compound_similarity`  # T4
 - `005_compound_clustering`  # T5
 - `006_compound_maximum_common_substructures`  # T6
-- `007_ligand_based_screening_machine_learning`  # T7
-- `008_pdb_protein_query`  # T8
+- `007_ligand_based_machine_learning`  # T7
+- `008_pdb_query`  # T8
 - `009_ligand_based_pharmacophores`  # T9
 - `010_binding_site_comparison`  # T10
 - `011_online_api_introduction`  # T11
```
Comment (jaimergp) [inline teachopencadd/README.md:16]: In hindsight, not all of those are APIs strictly speaking, so maybe we can instead use: 

```suggestion
- `011_online_api_webservices`  # T11
```

**Comment by jaimergp** [comment]
Yeah, I like it how they are! Anyway, as long as we provide these "per topic collections", the shortname of the folder/notebook is not _that_ important.

File: teachopencadd/README.md [hunk 2]
```diff
@@ -3,19 +3,19 @@
...
-- `004_ligand_based_screening_compound_similarity`  # T4
+- `001_chembl_query`  # T1
+- `002_compound_adme`  # T2
+- `003_compound_unwanted_substructures`  # T3
+- `004_compound_similarity`  # T4
 - `005_compound_clustering`  # T5
 - `006_compound_maximum_common_substructures`  # T6
-- `007_ligand_based_screening_machine_learning`  # T7
-- `008_pdb_protein_query`  # T8
+- `007_ligand_based_machine_learning`  # T7
```
Comment (AndreaVolkamer) [inline teachopencadd/README.md:12]:
```suggestion
- `007_compound_activity_prediction_machine_learning`  # T7
```

File: teachopencadd/README.md [hunk 3]
```diff
@@ -3,19 +3,19 @@
...
+- `008_pdb_query`  # T8
 - `009_ligand_based_pharmacophores`  # T9
```
Comment (AndreaVolkamer) [inline teachopencadd/README.md:14]:
```suggestion
- `009_compound_ensemble_pharmacophores`  # T9
```

**Comment by dominiquesydow** [comment]
> @dominiquesydow and @jaimergp just thought about the talktorial names, maybe we could come up with a more categorical name?
> Like the first ones operate all on compounds, so they could all start with compound_xxx, eg. change 001_chembl_comound_query to 01_compound_chembl_query, 004_ligand_based_screening_compound_similarity to 004_compound_similarity_ligand_based_screening and so on
Not sure if we can fit all, but maybe we find a good scheme?

__Point 1__
That goes again in the direction of what I first proposed, right? 
> Use foldername structure `index_biologicalentity_topic`
, where the biological entity is "compound" or "protein" and so forth?

Jaime argued against it because we will group the indices by category later anyways - and I agree that this scheme is too restricting. However, we could replace the remaining `ligand_based` with `compound`:
https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/README.md

__Point 2__
Regarding chembl/pdb/pubchem/..., I prefer the style `<index>_<database>_query` without specifying compound/protein/kinase instead of `<index>_<compound|protein|...>_<database>_query`.

@AndreaVolkamer, let me know what you think.

**Comment by dominiquesydow** [comment]
> I only suggested two more name changes. With that at least to me, they follow some kind of scheme +1
One could think of moving the word 'query' before the respective database query, so they are easier to identify as queries, but that's up to you.

You were faster :)
You were faster - ok will do

**Comment by AndreaVolkamer** [comment]
I think we were working on this at the same time. My Fazit with what we have we could have the following scheme:
`<index>_<compound|protein|binding site|protein_ligand|query>_<some more text>`

**Comment by dominiquesydow** [comment]
Then, we are almost back to were we started: 
https://github.com/volkamerlab/TeachOpenCADD/blob/98e2b0d1af3847c444bee196251f8c1f78c2b0ff/teachopencadd/README.md
> Use foldername structure index_biologicalentity_topic

Where Jaime argued against the following:
> I am not sure about categorizing stuff right away in the folder name. If we go this way, we should define a glossary of accepted biological entities before the list grows too big, but that might be too constraining and in the end we will have a category system that doesn't work or it's too vague.
> In my opinion, we should use titles that describe the contents precisely but succinctly (three, four words). Then, in the future, we ca use other means to provide collections of notebooks that make sense together (e.g. curated markdown files, symlinked directories, website sections...).

**Comment by dominiquesydow** [comment]
> I don't want to make this over complicated, since I agree with Jaime that we will group them later anyways, but it might just be easier to find things.

Ok, then I will apply the scheme where it fits easily - and omit the scheme where it is not helping. I think that's fair!

**Comment by dominiquesydow** [comment]
@AndreaVolkamer, in case you want a final look:
https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/README.md

Applied all your suggestions but kept `binding_site` (instead of the proposed `protein_binding_site`).

**Comment by AndreaVolkamer** [comment]
Looks great!

**Comment by dominiquesydow** [comment]
@jaimergp, @AndreaVolkamer, I moved the talktorials (except T8) as described in the PR description:

> Move talktorials to new folders; per talktorial do:
>    - Move notebook and rename to talktorial.ipynb.
>    - Move data (this included IO data files but also plots generated in the talktorial).
>    - Move images.
>    - Update IO file paths in talktorial.
>    - Remove talktorial numbers from notebook header (e.g. "# Talktorial 1"), since it is enough to have the index once stated in the talktorial folder name.
>    - Add note for talktorials 001-010: Reference to T1-T10 and the first TeachOpenCADD paper.

When T8 is merged (PR #32), I will apply these steps also there.
Let me know if you think of anything that is not as you expected.

**Comment by dominiquesydow** [comment]
Also, @jaimergp, @AndreaVolkamer, please also take a look at the PR question and let me know if you agree (a) with performing these tasks and (b) the proposed timeline:

> - [ ] Do this after refactoring for all notebooks (some notebooks are missing at the moment because they will come from an existing large notebook)?
>   - Fill talktorials' [`README`](https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/000_template/README.md): Automatically fetch notebook title, "Aims of this talktorial", and "Learning goals"?
>   - Fill talktorials' [`data/README`](https://github.com/volkamerlab/TeachOpenCADD/blob/renaming/teachopencadd/000_template/data/README.md): Automatically list here description of important IO/files (write `data/README` in notebook)?
>   - Add citation header to all notebooks (see (issue)[https://github.com/volkamerlab/TeachOpenCADD/issues/25])?

**Comment by dominiquesydow** [comment]
If filling `README` and `data/README` would work automatically, we could also add this before refactoring.

**Comment by dominiquesydow** [comment]
Note that `013_query_klifs` (will be split from T11a with parts from T12) and `014_binding_site_detection` (will be based on parts from T11a and T12) contain not talktorial, yet.

**Comment by dominiquesydow** [comment]
Two more comments on the `data/` folders

- We currently also write output plot files to this folder. This is not ideal but I would keep it this way to not over-complicate things.
- Example: Talktorial 003 needs data from 002. 
  - Currently, 003 fetches data from `002_.../data/`. 
  - I know we wanted stand-alone notebooks - I hesitated though to add the 002 output also to  `003_.../data/` because then it is no pipeline anymore. 
  - Suggestion (sticking to the example): We save the output from 002 to both `002_.../data/` and `003_.../data/` and offer in 003 two options: Run notebook from pipeline or standalone (with default standalone for CI)?

**Comment by jaimergp** [comment]
Wrt automating `README` generation, we'd need a script to parse the `*.ipynb` files and directories so these are generated. We could enforce this as part of the (future) CI, but for now it'd be a manual step we'd need to run before pushing (and this is easily forgettable). I am not against this at all (I'd love it!) but I'd say it's not a priority for these first steps. We can annotate the idea in an issue so we don't forget it for now.

In that direction we could use:

* https://github.com/rossant/ipymd
* At some point, [GH might support rendering `README.ipynb`](https://github.com/isaacs/github/issues/1425) directly.

***

> We currently also write output plot files to this `data`. This is not ideal but I would keep it this way to not over-complicate things.

No worries, that's fine.

> I know we wanted stand-alone notebooks

Yes, conceptually, but we will not be able to avoid some dependencies between them.

> because then it is no pipeline anymore.

Not really, right? If we are moving files from one step to another, it's a pipeline I think. I am not against this solution.

> We save the output from 002 to both 002_.../data/ and 003_.../data/

If 002 is run alone, the step that copies to 003 will fail or will be unnecessary. 

All in all, I think we don't need to overcomplicate this. While we are trying to make standalone notebooks whenever possible, some will have unavoidable dependencies. This can be stated at the top of the notebook in some kind of standardized way. As long as we freeze the notebook identifiers (`003_*`) and consider them immutable, references will work.

**Comment by dominiquesydow** [comment]
@jaimergp and @AndreaVolkamer - talktorial folders `008` and `010` (former T8 and T10) are updated now wrt the new folder/file scheme - thus we are ready to go! 

Since this whole renaming/file moving is quite tedious to review, we could merge now and start with packaging and refactoring. 

Since we do not merge back to `master` (but from `renaming` to `packaging`), I think it is ok to risk potential inconsistencies that I did not catch but will be caught during CI or later during refactoring.

Let me know what you think.

**Comment by dominiquesydow** [comment]
Each talktorial's folder README can be populated in a separate PR where we also explore the automatic options that @jaimergp pointed out.

**Comment by jaimergp** [comment]
Yeah, go ahead, no need to be super picky with reviews now, as this will be discovered during the CI phase! Thanks for the work here!

**Pull request merged**